snapshot/restore/restore.go

```diff
@@ -260,6 +270,54 @@ func (c *copier) copyDirectory(ctx context.Context, d fs.Directory, targetPath s
 	}), "copy directory contents")
 }
 
+func (c *copier) deleteExtraFilesInDir(ctx context.Context, d fs.Directory, targetPath string) error {
+	entries, err := fs.GetAllEntries(ctx, d)
+	if err != nil {
+		return errors.Wrap(err, "error reading directory")
+	}
+
+	// first classify snapshot entries to help with deletion (treat symlinks like normal files)
+	dirs := map[string]struct{}{}
+	files := map[string]struct{}{}
+
+	for _, e := range entries {
+		e := e
+
+		if e.IsDir() {
+			dirs[e.Name()] = struct{}{}
+		} else /* file */ {
+			files[e.Name()] = struct{}{}
+		}
+	}
+
+	// read existing entries on disk
+	existingEntries, err := os.ReadDir(targetPath)
```

Comment:
you should not assume targetPath is an operating system directory, there's `output` abstraction to support ZIP and TAR outputs too.

Reply:
thanks, wouldn't have thought of that!
Is typechecking c.Output the correct way?